### PR TITLE
Refine/tryit - added either/left/right types and type guards

### PR DESCRIPTION
**Pull request by stefaanv**
## Summary

Refines the `tryit` function by improving it's typings
`Left` (for failed) and `Right` (for successfull) types added
`Either` (Left or Right) and `Task` (async Either) types added
`isEither`, `isLeft` and `isRight` type guard functions added

This is the first of two PR's, in a second PR I would like to create a `pipe` function equivalent to `chain` but that also handles error states and both sync and async functions.

I ran into a problem with the `retry` tests because these use `try` with non-Error fault handling which I believe is not sound - I changes the tests.

## Related issue, if any:

None

## For any code change,
- [ ] Related documentation has been updated, if needed
- [x] Related tests have been added or updated, if needed
- [ ] Related benchmarks have been added or updated, if needed

## Does this PR introduce a breaking change?
Yes - `tryit()` will always return an error object if an `Error` is thrown in the function

## Bundle impact

_Calculating..._

### CODE REVIEW

**Comment by aleclarson** [comment]
Nice!

I think this one deserves a [Github discussion](https://github.com/orgs/radashi-org/discussions/new?category=ideas) as well. So we can explore the tradeoffs. 👍

File: src/async/tryit.ts [hunk 1]
```diff
@@ -1,46 +1,86 @@
...
+export type Task<T = unknown, E = Error> = Promise<Either<T, E>>
+
+/** syncEither or async Task */
+export type TaskOrEither<T, E> = T extends Promise<infer R>
+  ? Task<R, E>
+  : Either<T, E>
+
+export const isEither = (either: any): either is Either => {
+  return (
+    !isPromise(either) &&
```
Comment (aleclarson) [inline src/async/tryit.ts:22]: This `isPromise` check appears to be unnecessary

File: src/async/tryit.ts [hunk 2]
```diff
@@ -1,46 +1,86 @@
...
+    isArray(either) &&
+    either.length === 2 &&
+    (either[0] instanceof Error || either[0] === undefined)
+  )
+}
+
+/** returns `true` it the result is an error */
+export const isLeft = <T = unknown, E = Error>(
+  either: Either<T, E>,
+): either is Left<E> => isEither(either) && either[0] !== undefined
```
Comment (aleclarson) [inline src/async/tryit.ts:32]: Don't need the `isEither` check if `isLeft` requires an Either

File: src/async/tryit.ts [hunk 3]
```diff
@@ -1,46 +1,86 @@
...
+
+/** returns `true` it the result is NOT an error */
+export const isRight = <T = unknown, E = Error>(
+  either: Either<T, E>,
+): either is Right<T> => isEither(either) && either[0] === undefined
```
Comment (aleclarson) [inline src/async/tryit.ts:37]: Same here with the `isEither` check

File: src/async/tryit.ts [hunk 4]
```diff
@@ -1,46 +1,86 @@
...
+
+/** returns the error component of the either */
+export const left = <T = unknown, E = Error>(either: Either<T, E>): E =>
```
Comment (aleclarson) [inline src/async/tryit.ts:40]: Too much syntax sugar?

**Review by aleclarson — CHANGES_REQUESTED** [review]
- I'd prefer not to clog the `tryit.ts` module with a bunch of other stuff. Give each function its own module please.
- Move the isLeft/isRight functions to "typed" folder
- I don't think the syntax sugar afforded by `createLeft`/`createRight` are worth the abstraction
- The "Left" and "Right" lingo is probably more abstract than it should be. Why not "Fail" and "Pass"? Or maybe "Err" and "Ok" like Rust.
- "Either" could be renamed to "Result"
- isLeft -> isErr
- isRight -> isOk
- isEither -> isResult

**Comment by aleclarson** [comment]
I'm carrying this forward in #132 to be primarily type-based with a few helper functions like `isResult` and `toResult`.

**Comment by stefaanv** [comment]
Lots of comment - which is very positive.
As for the naming, I tried to be consistent with [ft-ts](https://gcanti.github.io/fp-ts/modules/Either.ts.html) and functional naming conventions.
Personally, I don't have a preference.  I'll be happy to folow your proposals.

**Comment by stefaanv** [comment]
As for the `isleft`/`isErr` and `isRight`/ `isOk` these can also be made into extension functions on the `Array` object.
I know this is opening a can of worms, is frowned upon, etc... because it alters the global context. but it does allow fluent style like
`result.isErr()`.  The chance that other libraries also use `isErr` and `isOk` to extend `Array` is not very big.

An alternative approach is to turn the Result into a class instead of an array, but that would be a rather bad breaking change IMO.

Let me know what you think.

**Comment by aleclarson** [comment]
> As for the `isleft`/`isErr` and `isRight`/ `isOk` these can also be made into extension functions on the `Array` object.
> I know this is opening a can of worms, is frowned upon, etc... because it alters the global context. but it does allow fluent style like
> `result.isErr()`. The chance that other libraries also use `isErr` and `isOk` to extend `Array` is not very big.

I'd be open to a separate `radashi-fluent` package on NPM. Its implementation can be kept in this repository under a `./fluent/` directory.



> An alternative approach is to turn the Result into a class instead of an array, but that would be a rather bad breaking change IMO.

I think arrays are seen as "lighter weight" at least syntactically. Not interested in deviating from whatever's most idiomatic in JavaScript. My mind remains open, nonetheless.

**Comment by aleclarson** [comment]
Closing in favor of #132

**Pull request closed without merging**